src/Memory/MemoryUtils.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "MemoryUtils.h"
+
+#include <absl/strings/numbers.h>
+#include <absl/strings/str_split.h>
+#include <stdlib.h>
+
+#include <string>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/ReadFileToString.h"
+#include "OrbitBase/Result.h"
+#include "OrbitBase/ThreadUtils.h"
+
+using orbit_grpc_protos::TotalMemoryUsageInfo;
+
+std::optional<TotalMemoryUsageInfo> ParseMemInfo(const std::string& meminfo_content) {
+  if (meminfo_content.empty()) return std::nullopt;
+
+  std::vector<std::string> lines = absl::StrSplit(meminfo_content, '\n');
+  if (lines.empty()) return std::nullopt;
+
+  constexpr int kMissingInfo = -1;
+  TotalMemoryUsageInfo memory_info;
+  memory_info.set_total_kb(kMissingInfo);
+  memory_info.set_free_kb(kMissingInfo);
+  memory_info.set_available_kb(kMissingInfo);
+  memory_info.set_buffers_kb(kMissingInfo);
+  memory_info.set_cached_kb(kMissingInfo);
+
+  constexpr int kNumLines = 5;
+  int mem_total, mem_free, mem_available, mem_buffers, mem_cached;
+  for (int i = 0, end = lines.size() > kNumLines ? kNumLines : lines.size(); i < end; ++i) {
+    const std::string& current_line = lines[i];
+    std::vector<std::string> splits = absl::StrSplit(current_line, ' ', absl::SkipWhitespace{});
+    if (splits.size() < 2) continue;
+    if (splits[0] == "MemTotal:" && absl::SimpleAtoi(splits[1], &mem_total)) {
+      memory_info.set_total_kb(mem_total);
+    } else if (splits[0] == "MemFree:" && absl::SimpleAtoi(splits[1], &mem_free)) {
+      memory_info.set_free_kb(mem_free);
+    } else if (splits[0] == "MemAvailable:" && absl::SimpleAtoi(splits[1], &mem_available)) {
+      memory_info.set_available_kb(mem_available);
+    } else if (splits[0] == "Buffers:" && absl::SimpleAtoi(splits[1], &mem_buffers)) {
+      memory_info.set_buffers_kb(mem_buffers);
+    } else if (splits[0] == "Cached:" && absl::SimpleAtoi(splits[1], &mem_cached)) {
+      memory_info.set_cached_kb(mem_cached);
+    }
+  }
+
+  return memory_info;
+}
+
+std::optional<TotalMemoryUsageInfo> GetTotalMemoryUsageInfo(uint64_t timestamp_ns) noexcept {
+  ErrorMessageOr<std::string> reading_result = orbit_base::ReadFileToString("/proc/meminfo");
```

Comment:
I just realized this information is not process specific, it is system-wide. is this intentional?

Reply:
(in which case pid does not make sense in proto, maybe rename proto to SystemMemoryUsage?)

Reply:
> I just realized this information is not process specific, it is system-wide. is this intentional?

Yes. We consider providing the system wide memory usage information at the first stage. 

Sorry about my misleading naming. I'm really not good at it :-)

Reply:
> (in which case pid does not make sense in proto, maybe rename proto to SystemMemoryUsage?)

Thanks for suggesting the name ` SystemMemoryUsage`!